File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/firebase/firestore/transactions/user/account/transaction/FirestorePurchaseHistory.java
package com.TritiumGaming.phasmophobiaevidencepicker.firebase.firestore.transactions.user.account.transaction;

import android.util.Log;

import com.TritiumGaming.phasmophobiaevidencepicker.firebase.firestore.listeners.OnFirestoreProcessListener;
import com.TritiumGaming.phasmophobiaevidencepicker.firebase.firestore.transactions.user.account.FirestoreTransactionHistory;
import com.google.firebase.Timestamp;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.SetOptions;

import java.util.HashMap;
import java.util.Map;

public class FirestorePurchaseHistory {

    private final static String COLLECTION_PURCHASE_HISTORY = "PurchaseHistory";
    private final static String DOCUMENT_PURCHASED_ITEM = "PurchaseItem";

    private final static String FIELD_DATE_PURCHASED = "datePurchased";

    private final static String
            FIELD_TYPE = "type",
            FIELD_PURCHASE_REFERENCE = "uuid";

    public static void init()
            throws Exception {

        getPurchaseDocument()
                .set(new HashMap<String, Object>(), SetOptions.merge())
                .addOnSuccessListener(unused ->
                        Log.d("Firestore", COLLECTION_PURCHASE_HISTORY + " successfully INITIALIZED!"))
                .addOnFailureListener(e -> {
                    Log.d("Firestore", COLLECTION_PURCHASE_HISTORY + "User Unlock History failed INITIALIZATION");
                    e.printStackTrace();
                })
                .addOnCompleteListener(task ->
                        Log.d("Firestore", COLLECTION_PURCHASE_HISTORY + "User Unlock History INITIALIZATION process complete!"));

    }

    public static CollectionReference getPurchaseHistoryCollection()
            throws Exception {

        return FirestoreTransactionHistory.getTransactionHistoryDocument().collection(COLLECTION_PURCHASE_HISTORY);
    }

    private static DocumentReference getPurchaseDocument()
            throws Exception {
        CollectionReference purchaseHistoryCollection = getPurchaseHistoryCollection();

        return purchaseHistoryCollection.document(DOCUMENT_PURCHASED_ITEM);
    }

    private static void addPurchaseDocument(
            String purchaseUUID, OnFirestoreProcessListener callback)
            throws Exception {
        CollectionReference purchaseHistoryCollection = getPurchaseHistoryCollection();

        if(purchaseUUID == null) { return; }

        Map<String, Object> documentData = new HashMap<>();
        documentData.put(FIELD_PURCHASE_REFERENCE, purchaseUUID);
        documentData.put(FIELD_DATE_PURCHASED, Timestamp.now());

        purchaseHistoryCollection.add(documentData)
                .addOnSuccessListener(unused -> {
                    callback.onSuccess();

                    Log.d("Firestore", "Purchase document of " + purchaseUUID + " GENERATED / LOCATED!");
                })
                .addOnFailureListener(e -> {
                    callback.onFailure();

                    Log.d("Firestore", "Purchase document of " + purchaseUUID + " could NOT be GENERATED / LOCATED!");
                    e.printStackTrace();
                })
                .addOnCompleteListener(task -> {
                    callback.onComplete();

                    Log.d("Firestore", "Purchase document of " + purchaseUUID + " process complete.");
                });
    }

    public static void purchaseCredits(String purchaseUUID, OnFirestoreProcessListener callback)
            throws Exception {
        addPurchaseDocument(purchaseUUID, callback);
    }

}

File: app-compose/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/data/firebase/firestore/transactions/user/account/transactions/types/FirestorePurchaseHistory.kt
package com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.account.transactions.types

import android.util.Log
import com.google.android.gms.tasks.Task
import com.google.firebase.Timestamp
import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.SetOptions
import com.tritiumgaming.phasmophobiaevidencepicker.data.firebase.firestore.transactions.user.account.transactions.FirestoreTransactionHistory.Companion.transactionHistoryDocument
import com.tritiumgaming.phasmophobiaevidencepicker.data.listeners.firestore.OnFirestoreProcessListener

class FirestorePurchaseHistory {

    companion object {
        private const val COLLECTION_PURCHASE_HISTORY = "PurchaseHistory"
        private const val DOCUMENT_PURCHASED_ITEM = "PurchaseItem"

        private const val FIELD_DATE_PURCHASED = "datePurchased"

        private const val FIELD_PURCHASE_REFERENCE = "product_ref"
        private const val FIELD_ORDER_ID = "order_id"

        @Throws(Exception::class)
        fun init() {
            purchaseDocument
                .set(HashMap<String, Any>(), SetOptions.merge())
                .addOnSuccessListener { unused: Void? ->
                    Log.d(
                        "Firestore",
                        "$COLLECTION_PURCHASE_HISTORY successfully INITIALIZED!"
                    )
                }
                .addOnFailureListener { e: Exception ->
                    Log.d(
                        "Firestore",
                        COLLECTION_PURCHASE_HISTORY + "User Unlock History failed INITIALIZATION"
                    )
                    e.printStackTrace()
                }
                .addOnCompleteListener { task: Task<Void?>? ->
                    Log.d(
                        "Firestore",
                        COLLECTION_PURCHASE_HISTORY + "User Unlock History INITIALIZATION process complete!"
                    )
                }
        }

        private val purchaseHistoryCollection: CollectionReference
            @Throws(Exception::class)
            get() = transactionHistoryDocument.collection(COLLECTION_PURCHASE_HISTORY)

        private val purchaseDocument: DocumentReference
            @Throws(Exception::class)
            get() {
                return purchaseHistoryCollection.document(DOCUMENT_PURCHASED_ITEM)
            }

        @Throws(Exception::class)
        fun addPurchaseDocument(
            purchaseReferenceDoc: DocumentReference?,
            orderID: String?,
            callback: OnFirestoreProcessListener?,
        ) {
            if (purchaseReferenceDoc == null || orderID == null) {
                return
            }

            val documentData: MutableMap<String, Any> = HashMap()
            documentData[FIELD_PURCHASE_REFERENCE] = purchaseReferenceDoc
            documentData[FIELD_ORDER_ID] = orderID
            documentData[FIELD_DATE_PURCHASED] = Timestamp.now()

            purchaseHistoryCollection.add(documentData)
                .addOnSuccessListener {
                    callback?.onSuccess()
                    Log.d(
                        "Firestore",
                        "Purchase document of ${ purchaseReferenceDoc.id } GENERATED / LOCATED!"
                    )
                }
                .addOnFailureListener { e: Exception ->
                    callback?.onFailure()
                    Log.d(
                        "Firestore",
                        "Purchase document of ${ purchaseReferenceDoc.id } could NOT be GENERATED / LOCATED!"
                    )
                    e.printStackTrace()
                }
                .addOnCompleteListener {
                    callback?.onComplete()
                    Log.d(
                        "Firestore",
                        "Purchase document of ${ purchaseReferenceDoc.id } process complete."
                    )
                }
        }
    }
}
